### PR TITLE
RemoteDocumentCache APIs for Indexing

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This changes the schema of the Remote Document Cache to rewrite the primary keys. The keys are now:

[ [prefix path ], collection group, read time, document ]

This allows for efficient scanning of indexed and non-indexed queries using `getAll()`. The PR also adds `getDocumentsFromCollectionGroup`, which means that all functionality required by Indexing is now available.

TODO:
- Benchmark the migration
- Re-write multi tab to use the collectionGroup+ReadTime index

### CODE REVIEW

File: packages/firestore/src/local/indexeddb_schema_converter.ts [hunk 1]
```diff
@@ -401,6 +423,60 @@ export class SchemaConverter implements SimpleDbSchemaConverter {
...
+    );
+    remoteDocumentStore.createIndex(
+      DbRemoteDocumentCollectionGroupIndex,
+      DbRemoteDocumentCollectionGroupIndexPath
+    );
+
+    const writes: Array<PersistencePromise<void>> = [];
+    return legacyRemoteDocumentStore
+      .iterate((_, legacyDocument) => {
+        const remoteDocumentStore = transaction.store<
```
Comment (wu-hui) [inline packages/firestore/src/local/indexeddb_schema_converter.ts:456]: Is this redeclaration of `remoteDocumentStore` necessary?

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts [hunk 2]
```diff
@@ -102,11 +104,17 @@ class IndexedDbRemoteDocumentCacheImpl implements IndexedDbRemoteDocumentCache {
...
-    documentKey: DocumentKey
+    documentKey: DocumentKey,
+    readTime: SnapshotVersion
   ): PersistencePromise<void> {
     const store = remoteDocumentsStore(transaction);
-    const key = dbKey(documentKey);
-    return store.delete(key);
+    const path = documentKey.path.toArray();
+    return store.delete([
+      path.slice(0, path.length - 2),
```
Comment (wu-hui) [inline packages/firestore/src/local/indexeddb_remote_document_cache.ts:113]: There are quite some places using `length - n` to get what we want, maybe we should put all these in `DocumentKey` instead.

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts [hunk 3]
```diff
@@ -247,55 +280,90 @@ class IndexedDbRemoteDocumentCacheImpl implements IndexedDbRemoteDocumentCache {
...
-      const collectionKey = collection.toArray();
-      const readTimeKey = toDbTimestampKey(sinceReadTime);
-      iterationOptions.range = IDBKeyRange.lowerBound(
-        [collectionKey, readTimeKey],
-        /* open= */ true
-      );
-      iterationOptions.index = DbRemoteDocumentCollectionReadTimeIndex;
-    }
+    const startKey = [
+      collection.popLast().toArray(),
```
Comment (wu-hui) [inline packages/firestore/src/local/indexeddb_remote_document_cache.ts:289]: Can we find a way to make it more readable, like signalling this is a `collectionPrefix` and below is `collectionId` somehow?

**Review by wu-hui — COMMENTED** [review]
I have not reviewed tests yet, but want to provide some feedback first.

Main concern is some places are not very readable. 

Also, it looks like we might need to do this for LevelDb as well?

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts (hunk 3 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/local/indexeddb_remote_document_cache.ts:284]: Added comments to the helpers.

File: packages/firestore/src/local/indexeddb_remote_document_cache.ts [hunk 4]
```diff
@@ -247,55 +280,90 @@ class IndexedDbRemoteDocumentCacheImpl implements IndexedDbRemoteDocumentCache {
...
+    offset: IndexOffset,
+    limit: number
+  ): PersistencePromise<MutableDocumentMap> {
+    debugAssert(limit > 0, 'Limit should be at least 1');
+
+    const path = offset.documentKey.path.toArray();
+    const startKey = [
+      collectionGroup,
+      toDbTimestampKey(offset.readTime),
+      path.length > 0 ? path.slice(0, path.length - 2) : [],
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/local/indexeddb_remote_document_cache.ts]: Added a helper.

File: packages/firestore/src/local/indexeddb_schema_converter.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/local/indexeddb_schema_converter.ts:440]: One of them is an IDBObjectStore, the other one is a SimpleDbObjectStore. I moved the object store creation out of this function.

File: .changeset/seven-dolphins-breathe.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+"@firebase/firestore": patch
+---
+
+The format of some of the IndexedDB data changed. This increases the performance of document lookups after an initial migration. If you do not want to migrate data, you can call `clearIndexedDbPersistence()` before invoking `enablIndexedDbPersistence()`.
```
Comment (markarndt) [inline .changeset/seven-dolphins-breathe.md:5]: Typo in line 5, needs another 'e', should be enableIndexedDbPersistence().

Comment (markarndt) [reply .changeset/seven-dolphins-breathe.md:5]: @egilmorez ... Talked to sebastian, and these changes are internal-only, no user-breaking effects. Thanks for adding me to the review.

@schmidt-sebastian ... Please feel free to include me in Firestore PRs, all platforms, especially the PRs that touch public-facing reference docs. Thanks!

**Review by wu-hui — APPROVED** [review]
LGTM.

**Pull request merged**